drivers/pinctrl/pinctrl_numicro.c

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022 SEAL AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/pinctrl/numicro-pinctrl.h>
+
+#define DT_DRV_COMPAT nuvoton_numicro_pinctrl
+
+#define REG_MFP(port, pin) (*(volatile uint32_t *)((uint32_t)DT_INST_REG_ADDR_BY_NAME(0, mfp) + \
+						   ((port) * 8) + \
+						   ((pin) > 7 ? 4 : 0)))
+
+#define REG_MFOS(port) (*(volatile uint32_t *)((uint32_t)DT_INST_REG_ADDR_BY_NAME(0, mfos) + \
+					       ((port) * 4)))
+
+#define MFP_CTL(pin, mfp) ((mfp) << (((pin) % 8) * 4))
+
+/**
+ * @brief Array containing pointers to each GPIO port.
+ *
+ * Entries will be NULL if the GPIO port is not enabled.
+ */
+static const struct device *const gpio_ports[] = {
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioa)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiob)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioc)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiod)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioe)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiof)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiog)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioh)),
+};
+
+/** Number of GPIO ports. */
+static const size_t gpio_ports_cnt = ARRAY_SIZE(gpio_ports);
+
+static int gpio_configure(const pinctrl_soc_pin_t *pin)
+{
+	const struct device *port;
+	gpio_flags_t flags = 0;
+
+	if (NUMICRO_PORT(pin->pinmux) >= gpio_ports_cnt) {
+		return -EINVAL;
+	}
+
+	port = gpio_ports[NUMICRO_PORT(pin->pinmux)];
+	if ((port == NULL) || (!device_is_ready(port))) {
+		return -ENODEV;
+	}
+
+	if (pin->pull_down != 0) {
+		flags |= GPIO_PULL_DOWN;
+	}
+
+	if (pin->pull_up != 0) {
+		flags |= GPIO_PULL_UP;
+	}
+
+	if (pin->open_drain != 0) {
+		flags |= GPIO_OPEN_DRAIN;
+	}
+
+	if (pin->schmitt_trigger != 0) {
+		flags |= NUMICRO_GPIO_INPUT_SCHMITT;
+	}
+
+	if (pin->input_disable == 0) {
+		flags |= GPIO_INPUT;
+	}
+
+	if (pin->input_debounce != 0) {
+		flags |= NUMICRO_GPIO_INPUT_DEBOUNCE;
+	}
+
+	return gpio_pin_configure(port, NUMICRO_PIN(pin->pinmux), flags);
```

Comment:
-1 on using GPIO API in pinctrl drivers, please, make them independent even if that involves certain code duplication. This has been the rule for all other drivers. Of course you can still use GPIO DT descriptions if needed.

Reply:
OK - pinctrl_stm32.c e.g. uses a separate function `gpio_stm32_configure` which is defined in gpio_stm32.c. Is that approach acceptable?

Reply:
No. That is the only case in-tree, mainly due to legacy reasons and it should be fixed at some point.